src/Analyzer/InDepthQueryTreeVisitor.h

```diff
@@ -84,4 +84,47 @@ class InDepthQueryTreeVisitor
 template <typename Derived>
 using ConstInDepthQueryTreeVisitor = InDepthQueryTreeVisitor<Derived, true /*const_visitor*/>;
 
+/** Visitor that use another visitor to visit node only if condition for visiting node is true.
+  * For example, your visitor need to visit only specific query tree nodes.
+  *
+  * Condition interface:
+  * struct Condition
+  * {
+  *     bool operator()(VisitQueryTreeNodeType & node)
+  *     {
+  *         return shouldNestedVisitorVisitNode(node);
+  *     }
+  * }
+  */
+template <typename Visitor, typename Condition, bool const_visitor = false>
+class InDepthQueryTreeConditionalVisitor : public InDepthQueryTreeVisitor<InDepthQueryTreeConditionalVisitor<Visitor, Condition, const_visitor>, const_visitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitor<InDepthQueryTreeConditionalVisitor<Visitor, Condition, const_visitor>, const_visitor>;
+    using VisitQueryTreeNodeType = typename Base::VisitQueryTreeNodeType;
+
+    explicit InDepthQueryTreeConditionalVisitor(Visitor & visitor_, Condition & condition_)
```

Comment:
Maybe it's better to make it "inheritable" instead of passing it as a constructor argument?